src/runtime/server/middleware/session/ipPinning.ts

```diff
@@ -0,0 +1,63 @@
+import { argon2id, hash, verify } from 'argon2'
+import { H3Event } from 'h3'
+import { useRuntimeConfig } from '#app'
+
+const argon2Options = {
+  // cryptographically-secure salt is generated automatically
+  type: argon2id, // resistant against GPU & tradeoff attacks
+  hashLength: 60
+}
+
+/**
+ * Get a hashed representation of the given IP address (for GDPR compliance)
+ * @param ip string|undefined The IP address to hash
+ */
+export const hashIpAddress = (ip: string | undefined): Promise<string | undefined> =>
+  !ip
+    ? Promise.resolve(undefined)
+    : hash(ip, argon2Options)
+
+/**
+ * Check that the given (raw) IP address and the hashed IP address match
+ * @param ip string|undefined The IP address to verify
+ * @param ipHash string|undefined The (hashed) IP address to test against
+ */
+export const ipAddressesMatch = (ip: string | undefined, ipHash: string | undefined): Promise<boolean> => !(ip && ipHash) ? Promise.resolve(false) : verify(ipHash, ip, argon2Options)
+
+/**
+ * Extract the IP address from an HTTP header
+ * @param header string|string[]|undefined The header value to inspect and extract the IP from
+ */
+const extractIpFromHeader = (header?: string | string[]): string | undefined => {
+  if (Array.isArray(header)) {
+    return header[0].split(',')[0]
+  }
+
+  if (typeof header === 'string') {
+    return header.split(',')[0]
+  }
+
+  return undefined
+}
+
+/**
+ * Get the IP address corresponding to the user's request
+ * @param event H3Event Event passing through middleware
+ */
+export const getRequestIpAddress = ({ req }: H3Event): string | undefined => {
+  const sessionOptions = useRuntimeConfig().session.session
+
+  // @ts-ignore TODO: Fix this `nuxi prepare` bug
+  const headerName = sessionOptions.ipPinning?.headerName
+
+  // @ts-ignore TODO: Fix this `nuxi prepare` bug
+  if (typeof sessionOptions.ipPinning === 'object' && 'headerName' in sessionOptions.ipPinning.headerName) {
+    return extractIpFromHeader(req.headers[headerName.toLowerCase()])
+  }
+
+  return req.connection?.remoteAddress ?? req.socket.remoteAddress
```

Comment:
You only mention `socket.remoteAddress` in the docs, what is `connection.remoteAddress` and is it preferred? Do you want to add it to the docs as well?

Reply:
Both `req.connection` and `req.socket` are `Socket` instances. However it seems to just be an alias